### ISSUE TITLE
PyTorch for quantum mechanics

### ISSUE THREAD

**Issue by PierreGuilmin** [pytorch/pytorch#82565]
### 🚀 The feature, motivation and pitch

As I begin working with PyTorch to simulate and optimize quantum systems, I propose to open this issue to list features that would be helpful.

Note that the goal is **not** to do quantum machine learning (e.g. https://www.tensorflow.org/quantum), but rather to simulate quantum system time evolution (using Schrödinger equation to begin with) and to perform gradient-based optimal control.

### Additional context

Related to #71446, but it is not active and specific to MIT's library.

_Don't hesitate to close this issue if you find it inappropriate, and prefer that I post individual feature requests._

**Comment by PierreGuilmin** [comment]
**Feature**
Compute the action of a sparse matrix exponential on a vector.

**Context**
The dynamic of a closed quantum system described by the quantum state $\ket{\psi(t)}$ is governed by Schrödinger equation

$$
i\hbar\frac{\partial}{\partial t}\ket{\psi(t)}=H\ket{\psi(t)}
$$

with $H$ an Hermitian operator. Given the state at time $t=0$ and if $H$ is independent of time, the unitary time evolution operator $U(t)=e^{-iHt/\hbar}$ gives the state at time $t$:

$$
\ket{\psi(t)}=U(t)\ket{\psi(0)}=e^{-iHt/\hbar}\ket{\psi(0)}
$$

Typically for all relevant physical problems, $H$ is a sparse matrix, so the main operation is the application of the exponential of $H$ on the vector $\ket{\psi(0)}$.

**References**
See [`scipy.sparse.linalg.expm_multiply`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.expm_multiply.html).

**Comment by jisaacso** [comment]
cc @cpuhrsch @bhosmer @jbschlosser on request for computing a sparse matrix exponential

**Comment by PierreGuilmin** [comment]
To foster the discussion, here's a reference from a group using TensorFlow to perform quantum optimal control. They too state that computing the unitary time evolution operator $U(t)=e^{-iHt/\hbar}$ is the main time bottleneck, and part 3 of the paper explains how they perform fast matrix exponentiation as well as fast gradient evaluation for this step, using relevant approximations.

Leung, N., Abdelhafez, M., Koch, J., & Schuster, D. (2017). Speedup for quantum optimal control from automatic differentiation based on graphics processing units. Physical Review A, 95(4), 042318.
https://journals.aps.org/pra/abstract/10.1103/PhysRevA.95.042318
